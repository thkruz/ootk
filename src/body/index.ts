export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
