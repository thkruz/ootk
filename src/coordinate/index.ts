export * from './Hill.js';
export * from './ClassicalElements.js';
export * from './EquinoctialElements.js';
export * from './FormatTle.js';
export * from './Geodetic.js';
export * from './ITRF.js';
export * from './J2000.js';
export * from './RelativeState.js';
export * from './RIC.js';
export * from './StateVector.js';
export * from './TEME.js';
export * from './Tle.js';
