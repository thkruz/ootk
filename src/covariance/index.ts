export * from './CovarianceSample.js';
export * from './StateCovariance.js';
