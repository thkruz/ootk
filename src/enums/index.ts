export { CatalogSource } from './CatalogSource.js';
export { CommLink } from './CommLink.js';
export { OrbitRegime } from './OrbitRegime.js';
export { Sgp4OpsMode } from './Sgp4OpsMode.js';
export { AngularDiameterMethod } from './AngularDiameterMethod.js';
export { AngularDistanceMethod } from './AngularDistanceMethod.js';
export { PassType } from './PassType.js';
