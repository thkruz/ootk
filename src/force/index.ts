export { EarthGravity } from './EarthGravity.js';
export { Force } from './Force.js';
export { ForceModel } from './ForceModel.js';
export { Gravity } from './Gravity.js';
export { SolarRadiationPressure } from './SolarRadiationPressure.js';
export { ThirdBodyGravity } from './ThirdBodyGravity.js';
export { Thrust } from './Thrust.js';

