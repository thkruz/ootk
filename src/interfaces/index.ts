export type { BaseObjectParams } from './BaseObjectParams.js';
export type { GroundPositionParams } from './GroundPositionParams.js';
export type { OptionsParams } from './OptionsParams.js';
export type { SatelliteParams } from './SatelliteParams.js';
export type { SensorParams } from './SensorParams.js';
export type { StarObjectParams } from './StarObjectParams.js';
export type { ClassicalElementsParams } from './ClassicalElementsParams.js';
export type { EquinoctialElementsParams } from './EquinoctialElementsParams.js';
export type { RfSensorParams } from './RfSensorParams.js';
