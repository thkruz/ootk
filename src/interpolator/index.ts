export { ChebyshevCoefficients } from './ChebyshevCoefficients.js';
export { ChebyshevInterpolator } from './ChebyshevInterpolator.js';
export { CubicSpline } from './CubicSpline.js';
export { CubicSplineInterpolator } from './CubicSplineInterpolator.js';
export { Interpolator } from './Interpolator.js';
export { LagrangeInterpolator } from './LagrangeInterpolator.js';
export { StateInterpolator } from './StateInterpolator.js';
export { VerletBlendInterpolator } from './VerletBlendInterpolator.js';

