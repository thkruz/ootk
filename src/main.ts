/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
