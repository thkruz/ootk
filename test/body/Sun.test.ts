import { Degrees, EpochUTC, Kilometers, Meters, Sun, Vector3D } from '../../src/main';

describe('Sun', () => {
  let exampleDate: Date;

  beforeEach(() => {
    const mockExampleDate = new Date(1705109326817);

    exampleDate = new Date(mockExampleDate.getTime());
  });

  /*
   * The 'azEl' method should return the azimuth and elevation of the sun given
   * a date, latitude, and longitude.
   */
  it('should return the azimuth and elevation of the sun given a date, latitude, and longitude', () => {
    const lat = 37.7749 as Degrees;
    const lon = -122.4194 as Degrees;
    const result = Sun.azEl(exampleDate, lat, lon);

    expect(result).toMatchSnapshot();
  });

  /*
   * The 'date2jSince2000' method should return the number of days since January
   * 1, 2000, for a given date.
   */
  it('should return the number of days since January 1, 2000, for a given date', () => {
    const result = Sun.date2jSince2000(exampleDate);

    expect(result).toMatchSnapshot();
  });

  /*
   * The 'diameter' method should return the angular diameter of the sun given
   * the observer's position and the sun's position.
   */
  it('should return the angular diameter of the sun given the observers position and the suns position', () => {
    const obsPos = new Vector3D<Kilometers>(0 as Kilometers, 0 as Kilometers, 0 as Kilometers);
    const sunPos = new Vector3D<Kilometers>(1 as Kilometers, 1 as Kilometers, 1 as Kilometers);
    const result = Sun.diameter(obsPos, sunPos);

    expect(result).toMatchSnapshot();
  });

  /*
   * The 'eclipseAngles' method should return the angles necessary to determine
   * if a satellite is in the shadow of the earth.
   */
  it('should return the angles necessary to determine if a satellite is in the shadow of the earth', () => {
    const satPos = new Vector3D<Kilometers>(0 as Kilometers, 0 as Kilometers, 0 as Kilometers);
    const sunPos = new Vector3D<Kilometers>(1 as Kilometers, 1 as Kilometers, 1 as Kilometers);
    const result = Sun.eclipseAngles(satPos, sunPos);

    expect(result).toMatchSnapshot();
  });

  /*
   * The 'eclipticLatitude' method should return the ecliptic latitude of the
   * sun given the solar latitude.
   */
  it('should return the ecliptic latitude of the sun given the solar latitude', () => {
    const B = 23.4397;
    const result = Sun.eclipticLatitude(B);

    expect(result).toMatchSnapshot();
  });

  /*
   * The 'eclipticLongitude' method should return the ecliptic longitude of the
   * sun given the solar mean anomaly.
   */
  it('should return the ecliptic longitude of the sun given the solar mean anomaly', () => {
    const M = 0;
    const result = Sun.eclipticLongitude(M);

    expect(result).toMatchSnapshot();
  });

  // The 'date2jSince2000' method should handle invalid dates.
  it('should handle invalid dates', () => {
    const date = new Date('invalid');
    const result = Sun.date2jSince2000(date);

    expect(result).toMatchSnapshot();
  });

  /*
   * The 'diameter' method should handle cases where the observer's position is
   * the same as the sun's position.
   */
  it('should handle cases where the observers position is the same as the suns position', () => {
    const obsPos = new Vector3D<Kilometers>(1 as Kilometers, 1 as Kilometers, 1 as Kilometers);
    const sunPos = new Vector3D<Kilometers>(1 as Kilometers, 1 as Kilometers, 1 as Kilometers);
    const result = Sun.diameter(obsPos, sunPos);

    expect(result).toMatchSnapshot();
  });

  /*
   * The 'eclipseAngles' method should handle cases where the satellite is
   * directly above or below the earth.
   */
  it('should handle cases where the satellite is directly above or below the earth', () => {
    const satPos = new Vector3D<Kilometers>(0 as Kilometers, 0 as Kilometers, 0 as Kilometers);
    const sunPos = new Vector3D<Kilometers>(0 as Kilometers, 0 as Kilometers, 1 as Kilometers);
    const result = Sun.eclipseAngles(satPos, sunPos);

    expect(result).toMatchSnapshot();
  });

  /*
   * The 'eclipticLatitude' method should handle extreme values of solar
   * latitude.
   */
  it('should handle extreme values of solar latitude', () => {
    const B = 90;
    const result = Sun.eclipticLatitude(B);

    expect(result).toMatchSnapshot();
  });

  /*
   * The 'eclipticLongitude' method should handle extreme values of solar mean
   * anomaly.
   */
  it('should handle extreme values of solar mean anomaly', () => {
    const M = 360;
    const result = Sun.eclipticLongitude(M);

    expect(result).toMatchSnapshot();
  });

  /*
   * The 'getSetJulian' method should return the Julian date of the sunrise,
   * sunset, or other solar event given the observer's position, date, and time.
   */
  it('should return the Julian date of the sunrise when the observers position, date, and time are provided', () => {
    // Mock input values
    const lat = 40;
    const lon = -75;
    const c = { ra: 0, dec: 0 };

    // Call the method
    const result = Sun.getSetJulian(0 as Meters, lon, lat, c.dec, 0, 0, 0);

    // Assert the result
    expect(result).toMatchSnapshot();
  });

  /*
   * The 'getSunTimeByAz' method should return the date and time of the sun's
   * position given a specific azimuth.
   */
  it('should return the date and time of the suns position given a specific azimuth', () => {
    const lat = 37.7749 as Degrees;
    const lon = -122.4194 as Degrees;
    const az = 1.5 as Degrees;
    const result = Sun.getSunTimeByAz(exampleDate, lat, lon, az);

    expect(result).toMatchSnapshot();
  });

  /*
   * The 'getTimes' method should return the sunrise, sunset, and other solar
   * event times for a given date, latitude, and longitude.
   */
  it('should return the sunrise, sunset, and other event times when given a date, latitude, and longitude', () => {
    // Mock input values
    const lat = 37.7749 as Degrees;
    const lon = -122.4194 as Degrees;

    // Call the getTimes method
    const result = Sun.getTimes(exampleDate, lat, lon);

    // Use toMatchSnapshot to test output of the getTimes method
    expect(result).toMatchSnapshot();
  });

  it('should return the same shadow result given the same time and position', () => {
    const result = Sun.shadow(
      EpochUTC.fromDateTime(exampleDate),
      new Vector3D<Kilometers>(8000 as Kilometers, 0 as Kilometers, 0 as Kilometers),
    );

    expect(result).toMatchSnapshot();
  });

  // positionApparent
  it('should return the same apparent position result given the same time and position', () => {
    const result = Sun.positionApparent(EpochUTC.fromDateTime(exampleDate));

    expect(result).toMatchSnapshot();
  });

  // sunlightLegacy
  it('should return the same sunlight result given the same positions', () => {
    const result = Sun.sunlightLegacy(new Vector3D(8000, 0, 0), new Vector3D(1000000, 0, 0));

    expect(result).toMatchSnapshot();
  });

  // lightingRatio
  it('should return the same lighting ratio result given the same positions', () => {
    const result = Sun.lightingRatio(
      new Vector3D<Kilometers>(8000 as Kilometers, 0 as Kilometers, 0 as Kilometers),
      new Vector3D<Kilometers>(1000000 as Kilometers, 0 as Kilometers, 0 as Kilometers),
    );

    expect(result).toMatchSnapshot();
  });

  // julianCyle
  it('should return the same julian cycle result given the same date', () => {
    const result = Sun.julianCycle(exampleDate, -75 as Degrees);

    expect(result).toMatchSnapshot();
  });
});
