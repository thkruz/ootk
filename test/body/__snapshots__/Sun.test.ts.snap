// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sun should handle cases where the observers position is the same as the suns position 1`] = `NaN`;

exports[`Sun should handle cases where the satellite is directly above or below the earth 1`] = `
Array [
  3.141592653589793,
  NaN,
  NaN,
]
`;

exports[`Sun should handle extreme values of solar latitude 1`] = `565.420893384895`;

exports[`Sun should handle extreme values of solar mean anomaly 1`] = `364.97003894087004`;

exports[`Sun should handle invalid dates 1`] = `NaN`;

exports[`Sun should return the Julian date of the sunrise when the observers position, date, and time are provided 1`] = `2451533.314279268`;

exports[`Sun should return the angles necessary to determine if a satellite is in the shadow of the earth 1`] = `
Array [
  3.141592653589793,
  NaN,
  NaN,
]
`;

exports[`Sun should return the angular diameter of the sun given the observers position and the suns position 1`] = `NaN`;

exports[`Sun should return the azimuth and elevation of the sun given a date, latitude, and longitude 1`] = `
Object {
  "az": 4.284718248237755,
  "el": -0.06852341194048539,
}
`;

exports[`Sun should return the date and time of the suns position given a specific azimuth 1`] = `2024-01-12T11:48:41.319Z`;

exports[`Sun should return the ecliptic latitude of the sun given the solar latitude 1`] = `147.22965595225526`;

exports[`Sun should return the ecliptic longitude of the sun given the solar mean anomaly 1`] = `4.938185716373701`;

exports[`Sun should return the number of days since January 1, 2000, for a given date 1`] = `8777.561652974691`;

exports[`Sun should return the same apparent position result given the same time and position 1`] = `
Vector3D {
  "x": 55101974.26909711,
  "y": -125165869.77224815,
  "z": -54258234.63367154,
}
`;

exports[`Sun should return the same julian cycle result given the same date 1`] = `8777`;

exports[`Sun should return the same lighting ratio result given the same positions 1`] = `1`;

exports[`Sun should return the same shadow result given the same time and position 1`] = `false`;

exports[`Sun should return the same sunlight result given the same positions 1`] = `1`;

exports[`Sun should return the sunrise, sunset, and other event times when given a date, latitude, and longitude 1`] = `
Object {
  "amateurDawn": 2024-01-12T14:08:48.295Z,
  "amateurDusk": 2024-01-13T02:29:02.928Z,
  "astronomicalDawn": 2024-01-12T13:53:05.375Z,
  "astronomicalDusk": 2024-01-13T02:44:45.848Z,
  "blueHourDawnEnd": 2024-01-12T15:08:10.931Z,
  "blueHourDawnStart": 2024-01-12T14:46:12.437Z,
  "blueHourDuskEnd": 2024-01-13T01:51:38.787Z,
  "blueHourDuskStart": 2024-01-13T01:29:40.292Z,
  "civilDawn": 2024-01-12T14:57:07.812Z,
  "civilDusk": 2024-01-13T01:40:43.411Z,
  "goldenHourDawnEnd": 2024-01-12T16:06:02.767Z,
  "goldenHourDawnStart": 2024-01-12T15:25:02.285Z,
  "goldenHourDuskEnd": 2024-01-13T01:12:48.939Z,
  "goldenHourDuskStart": 2024-01-13T00:31:48.456Z,
  "nadir": 2024-01-13T08:18:55.612Z,
  "nauticalDawn": 2024-01-12T14:24:41.646Z,
  "nauticalDusk": 2024-01-13T02:13:09.577Z,
  "solarNoon": 2024-01-12T20:18:55.612Z,
  "sunriseEnd": 2024-01-12T15:29:01.495Z,
  "sunriseStart": 2024-01-12T15:25:59.235Z,
  "sunsetEnd": 2024-01-13T01:11:51.988Z,
  "sunsetStart": 2024-01-13T01:08:49.728Z,
}
`;
